src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -327,7 +375,88 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             return csharpResponse.Result;
         }
 
-        private SemanticRange DataToSemanticRange(int lineDelta, int charDelta, int length, int tokenType, int tokenModifiers, SemanticRange? previousSemanticRange = null)
+        private async Task<SemanticTokens?> GetMatchingCSharpEditsResponseAsync(
+            TextDocumentIdentifier textDocumentIdentifier,
+            long? documentVersion,
+            string? previousResultId,
+            IReadOnlyList<int> previousCSharpTokens,
+            CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(previousResultId);
+
+            var parameter = new SemanticTokensDeltaParams
+            {
+                TextDocument = textDocumentIdentifier,
+                PreviousResultId = previousResultId,
+            };
+            var request = await _languageServer.SendRequestAsync(LanguageServerConstants.RazorProvideSemanticTokensEditsEndpoint, parameter);
+            var csharpResponse = await request.Returning<ProvideSemanticTokensResponse>(cancellationToken);
+
+            if (csharpResponse is null ||
+                (csharpResponse.HostDocumentSyncVersion != null && csharpResponse.HostDocumentSyncVersion != documentVersion))
+            {
+                // No C# response or C# is out of sync with us. Unrecoverable, return null to indicate no change. It will retry in a bit.
+                return null;
+            }
+
+            if (csharpResponse.Result != null)
+            {
+                // C#'s edit handler returned to us a full token set, so we don't need to do any additional work.
+                return csharpResponse.Result;
+            }
+
+            Assumes.NotNull(csharpResponse.EditResult);
+
+            if (!csharpResponse.EditResult.Edits.Any())
+            {
+                // If there aren't any edits, return the previous tokens with updated resultId.
+                return new SemanticTokens { Data = previousCSharpTokens.ToImmutableArray(), ResultId = csharpResponse.EditResult.ResultId };
```

Comment:
Personally I had assumed that if things were unchanged the same resultId would be returned, but on review I don't see that specifically stated in the LSP spec.

Reply:
Hmm, while we don't need to solve it here I DO think that ResultID's changing without code changes are a problem because of `_semanticTokensCache`.

If the ResultID changes without any code/project changes that means that each time VS asks for semantic tokens (last I checked it was every .5 seconds) we'll insert into the cache, meaning we'll get a lot of needless expanding and contracting of that cache. Fixing that in Razor without also fixing it in C# would be hard/pointless though. Does my logic check out?

Reply:
Ahh you bring up a great point! Yeah, thinking about it now, C# should probably just return the same resultId if the code is unchanged. Otherwise this introduces unnecessary insertions/deletions into the cache like you mentioned. I don't think this should be too difficult of a change, I can probably get a PR out for it today.